File: jsInterop.js
window.DNF = {
    history: {
        back() {
            history.back();
        },
        forward() {
            history.forward();
        },
        go(delta) {
            history.go(delta);
        },
        pushState(data, title, url) {
            history.pushState(data, title, location.pathname + url);
        },
        replaceState(data, title, url) {
            history.replaceState(data, title, location.pathname + url);
        }
    },
    toLocalhost(port) {
        if (/^https:\/\/ste80\.github\.io\/DotNetForce/.test(location.href)) {
            history.replaceState(null, null, location.href.replace(/^https:\/\/ste80\.github\.io\/DotNetForce/, `http://localhost:${port}`));
        }
    },
    getSessionStorage(key) {
        return sessionStorage.getItem(key);
    },
    setSessionStorage(key, value) {
        sessionStorage.setItem(key, value);
    },
    getLocalStorage(key) {
        return localStorage.getItem(key);
    },
    setLocalStorage(key, value) {
        localStorage.setItem(key, value);
    },
    addEventListener(type, callback) {
        addEventListener(type, function (ev) {
            var result = callback.invokeMethodAsync('Callback', type, JSON.stringify(ev));
            if (result === 'false') {
                return false;
            }
        });
    }
};

File: BlazorForce/JS/JsInterop.cs
using Newtonsoft.Json;
using System;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace BlazorForce
{
    public class JsInterop
    {
        public static readonly JsInterop Current = new JsInterop();

        public readonly HistoryJsInterop History = new HistoryJsInterop();

        public class HistoryJsInterop
        {
            public async Task BackAsync()
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.back");
            }
            public async Task ForwardAsync()
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.forward");
            }
            public async Task GoAsync(int? delta = null)
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.go", delta);
            }
            public async Task GoAsync(string delta)
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.go", delta);
            }
            public async Task PushStateAsync(object data, string title, string url)
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.pushState", data, title, url);
            }
            public async Task ReplaceStateAsync(object data, string title, string url)
            {
                await JSRuntime.Current.InvokeAsync<object>("DNF.history.replaceState", data, title, url);
            }
        }

        public readonly SessionStorageJsInterop SessionStorage = new SessionStorageJsInterop();

        public class SessionStorageJsInterop
        {
            public async Task<string> GetItemAsync(string key)
            {
                try
                {
                    return LZString.DecompressFromUTF16(await JSRuntime.Current.InvokeAsync<string>("DNF.sessionStorage.getItem", key));
                }
                catch { }
                return null;
            }
            public async Task SetItemAsync(string key, string value)
            {
                var compressed = LZString.CompressToUTF16(value);
                await JSRuntime.Current.InvokeAsync<object>("DNF.sessionStorage.setItem", key, compressed);
            }
            public async Task SetItemAsync(string key, object value)
            {
                var compressed = LZString.CompressToUTF16(JsonConvert.SerializeObject(value));
                await JSRuntime.Current.InvokeAsync<object>("DNF.sessionStorage.setItem", key, compressed);
            }
        }

        public readonly LocalStorageJsInterop LocalStorage = new LocalStorageJsInterop();

        public class LocalStorageJsInterop
        {
            public async Task<string> GetItemAsync(string key)
            {
                try
                {
                    return LZString.DecompressFromUTF16(await JSRuntime.Current.InvokeAsync<string>("DNF.localStorage.getItem", key));
                }
                catch { }
                return null;
            }
            public async Task SetItemAsync(string key, string value)
            {
                var compressed = LZString.CompressToUTF16(value);
                await JSRuntime.Current.InvokeAsync<object>("DNF.localStorage.setItem", key, compressed);
            }
            public async Task SetItemAsync(string key, object value)
            {
                var compressed = LZString.CompressToUTF16(JsonConvert.SerializeObject(value));
                await JSRuntime.Current.InvokeAsync<object>("DNF.localStorage.setItem", key, compressed);
            }
        }

        public async Task AddEventListenerAsync(string type)
        {
            await JSRuntime.Current.InvokeAsync<object>("DNF.addEventListener", type);
        }
    }
}